feat(docs): theming specification (#2296)
stardust-ui/react: docs/src/pages/ThemingSpecification.mdx

export const meta = {
  title: 'Theming specification',
}

## Goals and requirements

> The theming mechanism as part of Fluent UI should be flexible and provide support for as many scenarios as possible

Reviewer: This feels ambiguous, because "as possible" gives theming a lot of wiggle-room.
We should state that it must support the fluent theme, and all existing customization scenarios. It must also allow users to opt-out if the theming approach does not meet their needs.

##### Support for CSS stylesheets

Тhe users should be able to specify mappings from properties to classnames

**IMPORTANT**: applications which load their own CSS in other means should not pay the penalty of loading a runtime CSS-in-JS package, especially if the class names are known at time of authoring.

##### Support for CSSinJS

> CSSinJS support that will not tie the users to just one CSS in JS framework. The users should be able to define their styles and specify their engine that will be aligned to some defined interface.

Reviewer: Does this mean we will support multiple css in js frameworks with the fluent theme, or will just not prevent it because we also support passing classnames?

Author: I meant that we want to support multiple CSSinJS frameworks

In the CSS scenario, it is basically a way of allowing user to specify mapping to className for a new prop.

In the CSSinJS scenario, it is allowing user to have new properties (design terms) inside the styles functions.

#### Themes should cascade and merge allowing subtree themes to be updated at runtime and have the ability to adjust the incoming theme.

If some parameters/tokens are changed on the root of the application, the changes should be propagated to all nested subtrees.

#### RTL and disable animation modes should be configurable for the entire application or a subtree of the application.

It is important to allow the users to specify the RTL/LTR mode for the entire application once, but if necessary they can specify it only in a subtree of the application (for example one ChatMessage)

The same is true for the disable animations mode. It is important to allow the users to specify it for the entire application once, but if necessary they can specify it only in a subtree of the application (for example the `Loader` component should always be animated).

#### Parent components should be able to customize the styles for the child components

> The slots prop must be accessible in several means to ensure that a "recursive component" can be accurately styled.

Reviewer: Is a recursive component the only use-case? If so, it seems like there are several other ways to accomplish the same. (eg pass a prop to a child and have the child make decisions based on that prop.)

Author: Not really, below I have a details table with supporting parent styling slots/children. I will expand this to that scenario as well.

Author: I've updated this section. I understand that this can be achieved using different strategies, but we have the comparison in the table below anyway

The global tokens can be defined as a plain javascript object. We can offer some utilities for generating color palettes if there is an agreement of how those should be generated, but anyway, the global tokens should be defined by the users in any format they would like to structure them, and our framework should not depend on something that is defined (or not defined) there.

##### Component tokens

These are tokens which are specific for a component. They can use some of the global tokens for their values.

**Open for a debate here**

We have the components tokens in Teams theme and those were shown as a good abstraction. On the other hand, people overused them for adding inline overrides, because these tokens can be overridden on any instance of a component. So, maybe the best option will be to restrict the ways of where these tokens can be defined (only in the theme and not on instance level), so we can avoid similar misusages in future.

##### Component styles

The component styles should be functions that return CSSinJS definition that is compatible with the CSSinJS engine used. The input parameters of these functions are:

> - **design terms** - in the react implementation, these are basically subset of the component's properties. It is important to note here that for better performance we need to restrict these to be primitive values like: booleans, strings, numbers, so we can easily hash based on them.

Reviewer: > subset of the component's properties

Could you add some examples of what properties these are?

Author: Maybe the description is not the best. For the React components we can have many props. For example for a FluentButton we may have: onClick, content, icon, disabled, primary, circular etc. Out of these only subset of them are design terms: disabled, primary and circular.

- **design tokens** - these are basically the component's tokens. If we support inline overrides on these tokens, it will have some performance degradation

As we mention before, parents component should be able to specify styles for their slots. We can have two ways of how this can be implemented

- define slots' styles inside the styles of the parent component

```
const buttonStyles = {
  root: ({ terms, tokens }) => ({
    // some css in js definition
  }),
  icon: ({ terms, tokens }) => ({
    // some css in js definition
  }),
}
```

- create composed components for the slots if they require some additional styles

```
const buttonStyles = {
  root: ({ terms, tokens }) => ({
    // some css in js definition
  }),
}

const buttonIconStyles = {
  root: ({ terms, tokens }) => ({
    // some css in js definition
  }),
}
```

> There are props and cons to both approaches, so I will just try to list them all and we can discuss them:

Reviewer: This is a good list.

2 comments in favor of creating composed components:
- providing styling from a parent requires knowledge of the internals of another component, which creates coupling
- using a composed component makes it easy to swap in another component which might want all its own styles.

export const meta = {
  title: 'Theming specification',
}

## Goals and requirements

> 

Reviewer 2: I don't have an answer for this one, but it would be good to add a goal on bundle sizes for theming:

**App developers can get small bundles when only using a subset of components.**

I think the core difference between how Fabric and Stardust approached styling was this:

Fabric exports Checkbox and it comes pre-styled. That means if you only use the Checkbox in your app, you only get the Checkbox styles in your bundle; not the Button styles or anything else. You can define tokens in theme (colors, fonts, etc) which can get mapped into the styles. This means that bundle only contains applicable code, theme is small, and dev experience is nice (just use the Button.) You can even use Customizer to merge new styles on top of the defaults.

With the Stardust approach, styles come with the theme only. All components are "base" by default. Benefit being that you can completely hot swap the styling itself, the tech used to inject, tokens, anything, from the theme, without changing your app code. Unfortunately, decoupling this means you lose the treeshaking by default, where just using "uberTheme" with rendering the Button means you get the Button code, plus the styles for everything in uberTheme.

To make things smaller, you end up having to manually three shake your theme. It's possible to maybe do some webpack magic to shake out unused styling, but that would be major work and puts a pretty big requirement on the end user actually using webpack.

The "manual" tree shaking approach would be:

```tsx
import { CheckboxStyles } from '...';
const theme = createTheme(CheckboxStyles);

<Provider theme={theme}>
  <Checkbox/>
</Provider>
```

This gets messy because Checkbox actually depends on Icon. And ProviderBox. And Text. And other things. So CheckboxStyles better somehow pull those in too, because I had no idea until I scrubbed through the code.

I haven't come up with any other solutions than basically this:

We can provide a BaseButton. It doesn't care about stying.

For M365, we can provide a Button which does the Fabric thing for better tree shaking and better dev exp. Theme only has tokens, no styling.

Author: Thanks for bringing this up! Will add this in the goals section, this is really something that we need to address. The manual tree shaking was what I had in mind, but I see the point of knowing what the component depends on. Although this may seem like a very complicated matter, you still have to know this if you were styling this component, as you probably have some styles applied on the slots too... I am not sure that adding styles together with the component is the right things to do, as if we do that, the styles are no longer part of the theme, and we will break the concept of composing themes (like base Fluent -> Teams theme overrides) - I understand that we can have some mechanisms for adding style overrides, but I believe we will easily hit the wall where we will have styles across different component's version, and will be hard to keep track what's where if they are not organized in a centralized place. Anyway, I will add this in the goals section and we can iterate on it.

The theming mechanism as part of Fluent UI should be flexible and provide support for CSSinJS theme, and all existing customization scenarios. It must also allow users to opt-out if the theming approach does not meet their needs. This means that we need to have support for both CSS stylesheets and CSSinJS.

##### Support for CSS stylesheets

> Тhe users should be able to specify mappings from properties to classnames

Reviewer 2: I love this goal. Thanks for adding. It's important.

Just anecdotally; VSTS ended up with a forked version of Fabric; one of the reasons was because they wanted to use pure css for perf. We have ODSP saying they want 60ms initial page render times. Every single run time overhead contributor we can remove is great.

##### Variants

Variants are just mechanism for allowing users do add or remove some design terms from their components.

In order to enable this, the compose mechanism should provide to the users adding/removing some of the design terms. This can be easily done, by allowing the user to define mapping from props to design terms when they define the composed components.

```
export interface FluentButton extends ButtonProps {
  primary?: boolean
  circular?: boolean
}

// these code snippet is just illustrating a possible
// mechanism of how the design tokens can be created
// based on the props
const FluentButton = compose(Button, {
  displayName: 'FluentButton',
  mapPropsToTerms: props => ({
    primary: props.primary,
    circular: props.circular,
    hasIcon: !!props.icon,
  }),
})
```

##### Animations

Animation should be defined as part of the theme, but they should not be all loaded if they are not used. Child themes should be able to customize them.

```
// TBD
```

##### Icons

Icons should be defined as part of the theme, but they should not loaded if not used. Child themes should be able to customize them.

> 

Reviewer 2: Icons is a huge chuck of TeamsTheme. Are there ways to reduce the content? (I believe it's all the SVG content.) 

In Fabric we started with a HUGE icon set. I ended up separating the icons into chunks: one PLT1 chunk, and a number of other chunks which could be loaded on demand or deferred. This meant that some icons would be defined asynchronously. 

This would be something to consider here. Having icons be part of the theme would require a re-render of the entire tree in order to add more icons. (Maybe!)

Author: Yeah, I feel like we need to do a bit more investigation before deciding what is the best, that's why I didn't feel this section... We can iterate on it later.

Reviewer 3: Icons already created a lot of complexity and I am not sure that they are really part of the theme.

export const meta = {
  title: 'Theming specification',
}

## Goals and requirements

> The theming mechanism as part of Fluent UI should be flexible and provide support for CSSinJS theme, and all existing customization scenarios. It must also allow users to opt-out if the theming approach does not meet their needs. This means that we need to have support for both CSS stylesheets and CSSinJS.

Reviewer 3: ```suggestion
Components should have support for both CSS stylesheets and CSSinJS approaches. The theming mechanism as part of Fluent UI should be flexible and provide support for CSSinJS theme and all existing customization scenarios. Users can to opt-out if the theming approach does not meet their needs.
```

nit: I think that the meaningful part now is in the last sentence

##### Support for CSS stylesheets

> Тhe users should be able to specify mappings from properties to classnames

Reviewer 3: ```suggestion
Тhe users should be able to specify mappings from properties to classnames.
```

_nit_

**IMPORTANT**: applications which load their own CSS in other means should not pay the penalty of loading a runtime CSS-in-JS package, especially if the class names are known at time of authoring.

##### Support for CSSinJS

> CSSinJS support that will not tie the users to just one CSS in JS framework. The users should be able to define their styles and specify their engine that will be aligned to some defined interface.

Reviewer 3: We are not going to support any compatibility layer for different engines, i.e. if user decide to use JSS instead of Fela that is shipped by default user should to make his styles compatible on his side.

It will be great to emphasize this.

The client should be able to reuse the components between different themes. Supporting new theme in the application should not require building new components.

If new variants are added in the new theme, then the clients should be able to compose the existing component, by adding new design terms (via the props API). If the new theme doesn't need to support any new variant, nothing should be required from the client's side.

##### Variants

Variants are a way of allowing the users to specify new design terms for their components (among other customizations).

In the CSS scenario, it is basically a way of allowing user to specify mapping to className for a new prop.

In the CSSinJS scenario, it is allowing user to have new properties (design terms) inside the styles functions.

#### Themes should cascade and merge allowing subtree themes to be updated at runtime and have the ability to adjust the incoming theme.

If some parameters/tokens are changed on the root of the application, the changes should be propagated to all nested subtrees.

#### RTL and disable animation modes should be configurable for the entire application or a subtree of the application.

> It is important to allow the users to specify the RTL/LTR mode for the entire application once, but if necessary they can specify it only in a subtree of the application (for example one ChatMessage)

Reviewer 3: ```suggestion
It is important to allow the users to specify the RTL/LTR mode for the entire application once, but if necessary they can specify it only in a subtree of the application (for example one chat message).
```

_nit_

## Technical proposal

In the rest of the specification, there will be some proposals of how the themes, mappings and utilities can be structured.

#### CSS stylesheets usage

Users should be able to define mapping from props to classNames per component level.

This is inspired by the [React Semantic UI's className builders](https://github.com/Semantic-Org/Semantic-UI-React/blob/master/src/lib/classNameBuilders.js)

Here is an example of how the mappings for a FluentButton component can look like:

```
const fluentButtonMappings = props => {
  const { size, circular, attached, corner } = props
  return cx(
    size, // the value (small, medium, big)
    useKeyOnly(circular, 'circular'),
    useKeyOrValueAndKey(attached, 'attached'), // true or the value of the attached prop
    useValueAndKey(corner, 'corner'), // corner="left" will create the classNames `corner left`
  )
}

const FluentButton = compose(Button, {
  displayName: 'FluentButton',
  mapPropsToClasses: fluentButtonMappings,

> })

Reviewer 3: Not sure that this a good example as it hides complexity, i.e. how slots should work? Let's remove it for now.

Author: I've expanded the example to contain slots too. We can iterate on the actual API later. Is there a strong reason why we should remove it?

Reviewer 3: There is still one more confusing thing: `mapPropsToClasses: fluentButtonMappings,`.

- How it is different from `mapPropsToTerms`? I don't have design terms in CSS?
- How it is different from CSSinJS? In CSSinJS we are not coupling components with actual styles, but here we do it.

Author: I see what you're saying, maybe these should be somehow configured at a theme level. Will remove it for now.

```

#### CSSinJS usage

If users decide to use CSSinJS we have to propose some structure of how they can define all things required for theming their components.

> I will describe one proposal of how the theme object can look like:

Reviewer 3: Oops, as "I" a valid there?

```
export interface FluentButton extends ButtonProps {
  primary?: boolean
  circular?: boolean
}

// these code snippet is just illustrating a possible
// mechanism of how the design tokens can be created
// based on the props
const FluentButton = compose(Button, {
  displayName: 'FluentButton',
  mapPropsToTerms: props => ({
    primary: props.primary,
    circular: props.circular,
    hasIcon: !!props.icon,
  }),
})
```

##### Animations

Animation should be defined as part of the theme, but they should not be all loaded if they are not used. Child themes should be able to customize them.

```
// TBD
```

##### Icons

Icons should be defined as part of the theme, but they should not loaded if not used. Child themes should be able to customize them.

```
// TBD
```

##### Files and packaging

```
// themes/fluent/globalTokens.ts

const globalTokens = {
  colors: {
    brandForeground: 'white',
    brandBackground: 'red',

    defaultBackground: 'white',
    defaultForeground: 'black',
  }
}

> export default globalTokens

Reviewer 3: If we will use Styleguide it will be a list of separate constants:

```tsx
export colorBrandForeground = 'white'
export colorBrandBackground = 'red'
```

// themes/fluent/buttonTokens.ts

> const buttonTokens = globalTokens => ({

Reviewer 3: _I_ don't like this as it creates a hard coupling between non-treeshakable object `globalTokens` and `buttonTokens`, i.e. it's impossible for the bundler to understand that only `defaultBackground`, `defaultForeground`, `brandBackground` and `brandForeground` required there.

Author: What's the alternative proposal? We can import them directly in the `buttonTokens.ts`, but we want to update the tokens if they are updated in the the parent theme context. If we don't have the global tokens, than if we want to update some commonly used thing, we will need to update the tokens to multiple components...

Reviewer 3: If you remember https://codesandbox.io/s/t-rex-proto-1w8bp I had there a proposal to skip merging of tokens & styles. I **don't want** to say that we should go with it, I **want to say** that non-treeshakable tokens are not okay as they will create bloat for any styles:
```
buttonStyles() => (require) buttonTokens() => (require) globalTokens() 
```

Just to get current `siteVariables` I need 5kb of minified code:

![image](https://user-images.githubusercontent.com/14183168/73755881-020f9280-4767-11ea-99d7-a24986747ec2.png)

https://cdn.jsdelivr.net/combine/npm/@fluentui/react@0.43.1/dist/es/themes/teams/siteVariables.min.js,npm/@fluentui/react@0.43.1/dist/es/themes/teams/categoryColors.min.js,npm/@fluentui/react@0.43.1/dist/es/themes/teams/colors.min.js,npm/@fluentui/react@0.43.1/dist/es/utils/fontSizeUtility.min.js

***

However, with direct usages it can be better:

```js
// main.js
import {menuForeground } from './menu'

export const styles = () => ({
  color: menuForeground
})

// globals.js
export const colorBrandForeground = 'white'
export const colorBrandBackground = 'red'

// menu.js
import {colorBrandForeground, colorBrandBackground} from './globals'

export const menuForeground = colorBrandBackground
export const menuBackground = colorBrandForeground
```

Will be minified to:

```js
export default () => ({
    color: "red"
});
```

Demos:
https://rollupjs.org/repl/?version=1.31.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3Qm1lbnVGb3JlZ3JvdW5kJTIwJTdEJTIwZnJvbSUyMCcuJTJGbWVudSclNUNuJTVDbmV4cG9ydCUyMGNvbnN0JTIwc3R5bGVzJTIwJTNEJTIwKCklMjAlM0QlM0UlMjAoJTdCJTVDbiUyMCUyMGNvbG9yJTNBJTIwbWVudUZvcmVncm91bmQlNUNuJTdEKSU1Q24lNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyZ2xvYmFscy5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBjb25zdCUyMGNvbG9yQnJhbmRGb3JlZ3JvdW5kJTIwJTNEJTIwJ3doaXRlJyU1Q25leHBvcnQlMjBjb25zdCUyMGNvbG9yQnJhbmRCYWNrZ3JvdW5kJTIwJTNEJTIwJ3JlZCclMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIybWVudS5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAlN0Jjb2xvckJyYW5kRm9yZWdyb3VuZCUyQyUyMGNvbG9yQnJhbmRCYWNrZ3JvdW5kJTdEJTIwZnJvbSUyMCcuJTJGZ2xvYmFscyclNUNuJTVDbmV4cG9ydCUyMGNvbnN0JTIwbWVudUZvcmVncm91bmQlMjAlM0QlMjBjb2xvckJyYW5kQmFja2dyb3VuZCU1Q25leHBvcnQlMjBjb25zdCUyMG1lbnVCYWNrZ3JvdW5kJTIwJTNEJTIwY29sb3JCcmFuZEZvcmVncm91bmQlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXNtJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=
https://codesandbox.io/s/terser-el1eu

Author: Thanks a lot for sharing this @layershifter I remember this prototype and discussions, and from what I remember the scenario that this doesn't support is on the fly changes which needsto be propagated to the children themes. @levithomason  is this still a valid requirement?

The other thing is, theme composition will be a bit hard for managing if the tokens are not part of the theme. Once they become local imports, it will be hard to change anything or keep changes consistently across different themes. I am open for different proposals!

Author: After offline discussions, as we still don't have answer for how the things would be propagated down to the children theme if they are not in the context, we will leave this section as is, and in the future if we come up with better strategy, we will update it.
